### DISCUSSION TITLE
Proper way to source Discovery Data in Context, It

### DISCUSSION THREAD

**Discussion by TofuBug** [pester/Pester#2243] [Q&A] [answered]
Hello All,

I'm curious what the standard is when you have data from your BeforeDiscovery ScriptBlocks that aren't meant to be used as fuel for the -ForEach parameters but rather Generic shared data objects that might be reused across multiple Context including ones where -ForEach is being used.

Something like

```
BeforeDiscovery -ScriptBlock {
    $SharedData = @{
        TestPath = 'C:\Test'
        TestFilter = '*.txt'
    }

    $Tests = @('VM1','VM2','VM3'
}

InModuleScope -ModuleName SomeModule -Parameters @{
    SharedData = $SharedData
    Tests            = $Tests
} -ScriptBlock {
    Describe -Name 'Some Test' -Fixture {
        Context -Name 'Testing <VM> with <SharedData.TestPath>' -ForEach $Tests -Fixture {
            BeforeAll -ScriptBlock {
                $VM = $_
                $Path = $SharedData.TestPath
                $Filter = $SharedData.TestPath
            }
            It -Name 'Some Test with <VM>' -Test {}
        }
        Context -Name 'Testing with <SharedData.TestPath>' -Fixture {
            BeforeAll -ScriptBlock {
                $Path = $SharedData.TestPath
                $Filter = $SharedData.TestPath
            }
            It -Name 'Some <Path> with <Filter> -Test {}
        }
    }
}
```
Sorry for the contrived example but seems there should be some kind of parameter that like -TestCases and -ForEach allows you to pass a HashTable, or possibly array of objects and make it available across all cycles of that Context or Test something like -Data

If there already is a parameter that does this I'm missing I'm happy to be shown it.

I do understand I could have just duplicated the SharedData into my Tests

```
BeforeDiscovery -ScriptBlock {
    $Tests = @(
        @{
            VM = 'VM1' 
            SharedData = @{
                TestPath = 'C:\Test'
                TestFilter = '*.txt'
            }
        }
        @{
            VM = 'VM2' 
            SharedData = @{
                TestPath = 'C:\Test'
                TestFilter = '*.txt'
            }
        }
        @{
            VM = 'VM3' 
            SharedData = @{
                TestPath = 'C:\Test'
                TestFilter = '*.txt'
            }
        }
    }
}
```

but that just seems like a quick way for your test code to balloon out of control with anything beyond a trivial shared data like shown here.

I also know i can use the `$script:` scope to access these objects but that has ramifications if I have two Test files with the same `$script:` lines defining variables, when test are run independently everything is fine, but when run in an `Invoke-Pester where the entire collection of tests are run since discovery is run on all files before they execute the first test that defines that variable is now working off the last test that defined that scoped variable

Right now I'm solving the problem by something like

```
BeforeDiscovery -ScriptBlock {
    $SharedData = @{
        TestPath = 'C:\Test'
        TestFilter = '*.txt'
    }

    $Tests = @('VM1','VM2','VM3'
}

InModuleScope -ModuleName SomeModule -Parameters @{
    SharedData = $SharedData
    Tests            = $Tests
} -ScriptBlock {
    Describe -Name 'Some Test' -Fixture {
        Context -Name 'Testing <VM>' -ForEach $Tests -Fixture {
            BeforeAll -ScriptBlock {
                $VM = $_
            }
            Context -Name 'Using Shared Data <Path>' -ForEach $SharedData -Fixture {
                BeforeAll -ScriptBlock {
                    $SharedData = $_
                    $Path = $SharedData.TestPath
                    $Filter = $SharedData.TestPath
                }
                It -Name 'Some Test with <VM> and <Path>' -Test {}
        }
        Context -Name 'Testing with <Path>' -ForEach $SharedData -Fixture {
            BeforeAll -ScriptBlock {
                $SharedData = $_
                $Path = $SharedData.TestPath
                $Filter = $SharedData.TestPath
            }
            It -Name 'Some <Path> with <Filter> -Test {}
        }
    }
}
```

But that to me seems like I'm kind of bending the rules of what the -ForEach parameter is meant for

Maybe I'm overthinking it, just curious what other people do.

**Comment by fflaten** [comment, marked as answer]
If it's not used for generating tests then I'd place it in a root-level `BeforeAll`.

One limitation: You wouldn't be able to pass the variable to the `InModuleScope -Parameters` in your sample above since you've placed it around `Describe/Context` which makes it part of Discovery.

However, using `InModuleScope` like that is already discouraged in docs. If you moved `InModuleScope` inside `It/*Each/*All`, covering only the internal code that require it, then it would work as that code is executed in the Run-phase. 

Alternative:
You could also use param-block in the script and pass it using `New-PesterContainer -Data @{ ... }`, or in 5.3.0+ make it default parameter values. Script parameters are available in both Discovery and Run.
